### ISSUE TITLE
question about using tensorArray to do backpropagation

### ISSUE THREAD

**Issue by ximingwen** [tensorflow/tensorflow#59462]
I am confused about the tensorflow documentation here:

https://www.tensorflow.org/api_docs/python/tf/TensorArray:

Note that although the array can be read multiple times and positions can be overwritten, behavior may be undefined when storing multiple references to the same array and clear_after_read is False. In particular, avoid using methods like concat() to convert an intermediate TensorArray to a Tensor, then further modifying the TensorArray, particularly if you need to backprop through it later.

This is from tensorflow 2.11 which isn't included in tensorflow 2.7. So I want to know if we really need an intermediate Tensorarray and then concatenate it to a tensor and then calculate loss and do backpropogation, which way is suggested?

**Comment by MathurUtkarsh** [comment]
Issue #59462

The suggested approach would be to avoid using TensorArray and concatenating it to a Tensor, if you are using TensorFlow 2.7. Instead, you can use other TensorFlow operations to perform the same operations without the need for TensorArray. 

For example, you can use the tf.stack() function to concatenate multiple tensors along a given axis. You can also use the tf.data API to create a dataset and use the dataset to feed the data to the model.

When it comes to calculate the loss, you can use the loss functions provided by TensorFlow such as tf.losses.mean_squared_error and tf.losses.categorical_crossentropy to calculate the loss and backpropagate through it.

In summary, to avoid using TensorArray you can use the combination of tf.stack(), tf.data API and TensorFlow's loss functions to perform the same operations in TensorFlow 2.7.

Documentation - https://www.tensorflow.org/api_docs/python/tf/TensorArray

**Comment by MathurUtkarsh** [comment]
Assign this issue to me.

**Comment by MathurUtkarsh** [comment]
Issue #59462

The suggested approach would be to avoid using TensorArray and concatenating it to a Tensor, if you are using TensorFlow 2.7. Instead, you can use other TensorFlow operations to perform the same operations without the need for TensorArray.

For example, you can use the tf.stack() function to concatenate multiple tensors along a given axis. You can also use the tf.data API to create a dataset and use the dataset to feed the data to the model.

When it comes to calculate the loss, you can use the loss functions provided by TensorFlow such as tf.losses.mean_squared_error and tf.losses.categorical_crossentropy to calculate the loss and backpropagate through it.

In summary, to avoid using TensorArray you can use the combination of tf.stack(), tf.data API and TensorFlow's loss functions to perform the same operations in TensorFlow 2.7.

**Comment by indrasn0wal** [comment]
I would like to work on this issue.

**Comment by sushreebarsa** [comment]
@ximingwen tf.TensorArray class is meant to be used with dynamic iteration primitives such as while_loop and map_fn. It supports gradient back-propagation via special "flow" control flow dependencies. There are some other APIs that can be used instead such as tf. stack() and tf.data in TF latest versions. Please let us know if it helps?
Thank you!

**Comment by ximingwen** [comment]
> @ximingwen tf.TensorArray class is meant to be used with dynamic iteration primitives such as while_loop and map_fn. It supports gradient back-propagation via special "flow" control flow dependencies. There are some other APIs that can be used instead such as tf. stack() and tf.data in TF latest versions. Please let us know if it helps? Thank you!



> Issue #59462
> 
> The suggested approach would be to avoid using TensorArray and concatenating it to a Tensor, if you are using TensorFlow 2.7. Instead, you can use other TensorFlow operations to perform the same operations without the need for TensorArray.
> 
> For example, you can use the tf.stack() function to concatenate multiple tensors along a given axis. You can also use the tf.data API to create a dataset and use the dataset to feed the data to the model.
> 
> When it comes to calculate the loss, you can use the loss functions provided by TensorFlow such as tf.losses.mean_squared_error and tf.losses.categorical_crossentropy to calculate the loss and backpropagate through it.
> 
> In summary, to avoid using TensorArray you can use the combination of tf.stack(), tf.data API and TensorFlow's loss functions to perform the same operations in TensorFlow 2.7.

Hi, Thank you very much for the reply. However, when training Tensorflow use strategy, it is suggested that we wrap the strategy.run under tf.function. In this way,  it is also suggested that not replying on python's side effect like appending. Then if I really need to do something like tf.stack(list), which list needs to generate dynamically for each call, which way should be the best way to go?
